### PR TITLE
fix read buffer usage

### PR DESCRIPTION
**Pull request by clockfort**
The reflection is temporary, I'm adding this ability to Thrift. It's useful for anyone pooling Thrift connections.

I wrote a test for this and used it to develop on locally, but it's pretty impractical as a real test on check-in, so I did not include it. The test was:
```
package com.palantir.atlasdb.keyvalue.cassandra;

import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.net.InetSocketAddress;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSortedSet;
import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfigManager;
import com.palantir.atlasdb.cassandra.ImmutableCassandraKeyValueServiceConfig;
import com.palantir.atlasdb.encoding.PtBytes;
import com.palantir.atlasdb.keyvalue.api.Cell;
import com.palantir.atlasdb.keyvalue.api.ColumnSelection;
import com.palantir.atlasdb.keyvalue.api.TableReference;
import com.palantir.atlasdb.protos.generated.TableMetadataPersistence;
import com.palantir.atlasdb.table.description.TableDefinition;
import com.palantir.atlasdb.table.description.ValueType;
import com.palantir.atlasdb.transaction.api.ConflictHandler;
import com.sun.management.HotSpotDiagnosticMXBean;

public class TestWriteBufferUsageInCKVS {

    public static void main(String[] args) throws IOException, InterruptedException {
        // setup a basic Cassandra-backed Atlas
        CassandraKeyValueService kvs = CassandraKeyValueService.create(
                CassandraKeyValueServiceConfigManager.createSimpleManager(
                        ImmutableCassandraKeyValueServiceConfig.builder()
                                .addServers(new InetSocketAddress("localhost", 9160))
                                .keyspace("atlas")
                                .ssl(false)
                                .replicationFactor(1)
                                .build()));

        TableReference blobTable = TableReference.createFromFullyQualifiedName("example.blob_table");
        kvs.createTable(
                blobTable,
                new TableDefinition() {{
                    rowName();
                    rowComponent("key", ValueType.FIXED_LONG);
                    columns();
                    column("blob", "blob", ValueType.BLOB);
                    conflictHandler(ConflictHandler.IGNORE_ALL);
                    sweepStrategy(TableMetadataPersistence.SweepStrategy.NOTHING);
                }}.toTableMetadata().persistToBytes()
        );

        kvs.truncateTable(blobTable);

        ExecutorService writeExecutor = Executors.newFixedThreadPool(64);
        for (int i=0; i < 200; i++) {
            final int j = i;
            writeExecutor.submit((Callable<Void>) () -> {
                byte[] largeValue = new byte[15*1024*1024];
                ThreadLocalRandom.current().nextBytes(largeValue);
                kvs.put(blobTable,
                        ImmutableMap.of(Cell.create(ValueType.FIXED_LONG.convertFromJava(Long.valueOf(j)), "blob".getBytes()), largeValue), 1);
                System.out.println("finished " + j);
                return null;
            });
        }
        writeExecutor.shutdown();
        writeExecutor.awaitTermination(5, TimeUnit.MINUTES);

        dumpHeap("heapDumpWithPossiblyManyWriteBuffers.hprof");

        ExecutorService readExecutor = Executors.newFixedThreadPool(64);
        for (int i=0; i < 200; i++) {
            final int j = i;
            readExecutor.submit((Callable<Void>) () -> {
                kvs.getRows(blobTable, ImmutableSortedSet.<byte[]>orderedBy(PtBytes.BYTES_COMPARATOR).add(ValueType.FIXED_LONG.convertFromJava(Long.valueOf(j))).build(), ColumnSelection.all(), 1);
                System.out.println("finished " + j);
                return null;
            });
        }
        readExecutor.shutdown();
        readExecutor.awaitTermination(5, TimeUnit.MINUTES);

        dumpHeap("heapDumpWithPossiblyManyReadBuffers.hprof");

        kvs.truncateTable(blobTable);
    }

    private static void dumpHeap(String fileName) throws IOException {
        ManagementFactory.newPlatformMXBeanProxy(ManagementFactory.getPlatformMBeanServer(),
                "com.sun.management:type=HotSpotDiagnostic", HotSpotDiagnosticMXBean.class).dumpHeap(fileName, true);
    }
}

```

The 20,000 foot view of the test results before and after the change speak for themselves:
```
[work atlasdb]$ ls -lha *hprof
-rw-------. 1 clockfort clockfort 968M Jun 14 16:18 heapDumpWithPossiblyManyReadBuffers.hprof
-rw-------. 1 clockfort clockfort 8.3M Jun 14 16:17 heapDumpWithPossiblyManyWriteBuffers.hprof
[work atlasdb]$ rm *hprof
[work atlasdb]$ ls -lha *hprof
-rw-------. 1 clockfort clockfort 8.8M Jun 14 16:48 heapDumpWithPossiblyManyReadBuffers.hprof
-rw-------. 1 clockfort clockfort 8.3M Jun 14 16:48 heapDumpWithPossiblyManyWriteBuffers.hprof
```

( @schlosna this also answers your previous question re: cleaning up other types of buffers, which I investigated both in testing and read the code for now; this is the only thing we need to clean up )

### CODE REVIEW

**Comment by tmgordeeva** [comment]
was worried about nulls but clockfort convinced me we don't need more null checking

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 1]
```diff
@@ -110,13 +112,20 @@ protected int getPoolSize() {
...
+                            readBuffer_.setAccessible(true);
+                            TMemoryInputTransport tMemoryInputTransport = (TMemoryInputTransport) readBuffer_.get(transport);
+                            byte[] underlyingBuffer = tMemoryInputTransport.getBuffer();
+                            if (underlyingBuffer != null) {
+                                log.debug("During {} check-in, cleaned up a read buffer of {} bytes", resource, underlyingBuffer.length);
+                                tMemoryInputTransport.clear();
+                            }
                         }
+                    } catch (Exception e) {
+                        log.debug("Couldn't clean up read buffers on pool check-in.", e);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: nit: probably worth extracting this out to a separate method.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 2]
```diff
@@ -110,13 +112,20 @@ protected int getPoolSize() {
...
+                    try {
+                        TTransport transport = resource.getInputProtocol().getTransport();
+                        if (transport instanceof TFramedTransport) {
+                            Field readBuffer_ = ((TFramedTransport) transport).getClass().getDeclaredField("readBuffer_");
+                            readBuffer_.setAccessible(true);
+                            TMemoryInputTransport tMemoryInputTransport = (TMemoryInputTransport) readBuffer_.get(transport);
+                            byte[] underlyingBuffer = tMemoryInputTransport.getBuffer();
+                            if (underlyingBuffer != null) {
+                                log.debug("During {} check-in, cleaned up a read buffer of {} bytes", resource, underlyingBuffer.length);
+                                tMemoryInputTransport.clear();
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: I realize upstream Thrift is unlikely to change, but would prefer putting `tMemoryInputTransport.clear();` outside the `if` to always clear.

**Comment by schlosna** [comment]
a couple nits 👍

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java (hunk 2 above)
Comment (clockfort) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: If they move away from TFramedTransport there is no way any of this stuff will have the right variable names, etc anyways

**Pull request merged**